pkg/util/bandwidth/linux.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bandwidth
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/exec"
+
+	"github.com/golang/glog"
+)
+
+// tcShaper provides an implementation of the BandwidthShaper interface on Linux using the 'tc' tool.
+// In general, using this requires that the caller posses the NET_CAP_ADMIN capability, though if you
+// do this within an container, it only requires the NS_CAPABLE capability for manipulations to that
+// container's network namespace.
+// Uses the hierarchical token bucket queueing discipline (htb), this requires Linux 2.4.20 or newer
+// or a custom kernel with that queueing discipline backported.
+
+type tcShaper struct {
+	e exec.Interface
+}
+
+func NewTCShaper() BandwidthShaper {
+	return &tcShaper{e: exec.New()}
+}
+
+func (t *tcShaper) execAndLogString(cmd string) error {
```

Comment:
I did all the iptables work - I know.

On Thu, Aug 6, 2015 at 9:46 PM, Brendan Burns <notifications@github.com>
wrote:

> In pkg/util/bandwidth/linux.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11988#discussion_r36490750>
> :
>
> > +// tcShaper provides an implementation of the BandwidthShaper interface on Linux using the 'tc' tool.
> > +// In general, using this requires that the caller posses the NET_CAP_ADMIN capability, though if you
> > +// do this within an container, it only requires the NS_CAPABLE capability for manipulations to that
> > +// container's network namespace.
> > +// Uses the hierarchical token bucket queueing discipline (htb), this requires Linux 2.4.20 or newer
> > +// or a custom kernel with that queueing discipline backported.
> > +
> > +type tcShaper struct {
> > +	e exec.Interface
> > +}
> > +
> > +func NewTCShaper() BandwidthShaper {
> > +	return &tcShaper{e: exec.New()}
> > +}
> > +
> > +func (t *tcShaper) execAndLogString(cmd string) error {
>
> fwiw, it makes it a lot harder to cut and paste them to test them on the
> command line...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11988/files#r36490750>
> .
>